### PR TITLE
Lab02 - Bubble and insertion sort with unit tests

### LINKED ISSUE #253: Lab02 - Bubble and Insertion Sort
- Complete bubble sort algorithm 
- Complete insertion sort algorithm
- Complete test cases for both sort algortihm

### PR DESCRIPTION
**Pull request by willshen8**
Fixes #253 

Review of colleague's PR #252 

#### Changes proposed in this PR:

- Bubble sort
- Insertion sort
- Test cases for both sort algorithm with 100% coverage

### CODE REVIEW

File: 02_bubble/willshen8/bubble_test.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		{name: "Already sorted slice", input: []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10},
+			expected: []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}},
+		{name: "Slice with one number different to rest", input: []int{1, 1, 1, 2, 1},
+			expected: []int{1, 1, 1, 1, 2}},
+		{name: "Slice with negative number", input: []int{1, -1}, expected: []int{-1, 1}},
+	}
+
+	for _, test := range tests {
+		actual := insertionSort(test.input)
+		testResult := Equal(actual, test.expected)
```
Comment (golangcibot) [inline 02_bubble/willshen8/bubble_test.go]: undeclared name: `Equal` (from `typecheck`)

File: 02_bubble/willshen8/bubble.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+			sliceCopy[emptySlot] = sliceCopy[emptySlot-1]
+			emptySlot--
+		}
+		sliceCopy[emptySlot] = currValue
+
+	}
+	return sliceCopy
+}
+
+func swap(slice []int, pos1 int, pos2 int) []int {
```
Comment (marksomething) [inline 02_bubble/willshen8/bubble.go]: Go allows you to do in-place swaps `slice[pos1], slice[pos2] = slice[pos2], slice[pos1]`
You can replace this function with that.

File: 02_bubble/willshen8/bubble.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	return sliceCopy
+}
+
+func insertionSort(s []int) []int {
+	sliceCopy := append(s[:0:0], s...)
+	for i := 1; i < len(s); i++ {
+		currValue := sliceCopy[i]
+		emptySlot := i
+
+		for emptySlot > 0 && sliceCopy[emptySlot-1] > currValue {
```
Comment (marksomething) [inline 02_bubble/willshen8/bubble.go]: an alternative here is to use copy() to move the existing entries rather than shuffling them along one at a time.

File: 02_bubble/willshen8/bubble.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	for !sorted {
+		swapOccurred := false
+		for i := 0; i < sliceLength-1; i++ {
+			if sliceCopy[i+1] < sliceCopy[i] {
+				swap(sliceCopy, i, i+1)
+				swapOccurred = true
+			}
+		}
+		if !swapOccurred {
+			sorted = true
```
Comment (marksomething) [inline 02_bubble/willshen8/bubble.go:25]: Had a review comment on my code suggest moving the `return` here. It's kinda neat and removes the need for the `sorted` variable.

File: 02_bubble/willshen8/bubble.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+		}
+	}
+	return sliceCopy
+}
+
+func insertionSort(s []int) []int {
+	sliceCopy := append(s[:0:0], s...)
+	for i := 1; i < len(s); i++ {
```
Comment (marksomething) [inline 02_bubble/willshen8/bubble.go]: Looping an entire slice is more idiomatically done using `range`. You have used that technique elsewhere.

File: 02_bubble/willshen8/bubble.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func swap(slice []int, pos1 int, pos2 int) []int {
+	temp := slice[pos2]
+	slice[pos2] = slice[pos1]
+	slice[pos1] = temp
+	return slice
+}
+
+func equal(a, b []int) bool {
```
Comment (marksomething) [inline 02_bubble/willshen8/bubble.go:45]: Have a look at testify/assert, it was recommended to me by @juliaogris and spares you from needing to implement your own slice equality (and diff printing) code.

File: 02_bubble/willshen8/bubble_test.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	for _, test := range tests {
+		t.Logf("Running test: %v", test.name)
+		actual := equal(test.sliceA, test.sliceB)
+		if actual != test.expected {
+			t.Fatalf("Test failed! Expected: %t, Got: %t", test.expected, actual)
+		}
+	}
+}
+
+func TestInsertionSort(t *testing.T) {
```
Comment (marksomething) [inline 02_bubble/willshen8/bubble_test.go]: Consider using the same cases between your two sorting implementations. These are all really good choices for test cases 👍

File: 02_bubble/willshen8/bubble_test.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+	tests := []struct {
```
Comment (marksomething) [inline 02_bubble/willshen8/bubble_test.go]: Consider using a `map[string]struct{[]int, []int}`

File: 02_bubble/willshen8/bubble_test.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+	}{
+		{name: "Empty slice", input: []int{}, expected: []int{}},
+		{name: "Reverse sorted slice", input: []int{10, 9, 8, 7, 6, 5, 4, 3, 2, 1},
+			expected: []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}},
+		{name: "Already sorted slice", input: []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10},
+			expected: []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}},
+		{name: "Slice with negative number", input: []int{1, -1}, expected: []int{-1, 1}},
+	}
+
+	for _, test := range tests {
```
Comment (marksomething) [inline 02_bubble/willshen8/bubble_test.go:38]: Have a look at "Introducing sub tests" in [this blog](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests)

**Review by marksomething — CHANGES_REQUESTED** [review]
Nice work! You code is really good in general, most of the ideas here I only learnt from other peoples code reviews :)

File: 02_bubble/willshen8/bubble.go (hunk 2 above)
Comment (willshen8) [reply 02_bubble/willshen8/bubble.go]: Thanks, done.

File: 02_bubble/willshen8/bubble.go (hunk 5 above)
Comment (willshen8) [reply 02_bubble/willshen8/bubble.go]: Thanks, just very used to this kind of loop and forgot to use the new style.

File: 02_bubble/willshen8/bubble_test.go (hunk 7 above)
Comment (willshen8) [reply 02_bubble/willshen8/bubble_test.go]: Sorry, not sure what you meant here.

File: 02_bubble/willshen8/bubble.go (hunk 6 above)
Comment (willshen8) [reply 02_bubble/willshen8/bubble.go:45]: Great point, that's for pointing that out for me. Will use testify next time.

File: 02_bubble/willshen8/bubble_test.go (hunk 8 above)
Comment (willshen8) [reply 02_bubble/willshen8/bubble_test.go]: Map actually is slower as it iterate through both the key and value. So in this case the original struct is better in performance.

File: 02_bubble/willshen8/bubble.go (hunk 2 above)
Comment (anzboi) [reply 02_bubble/willshen8/bubble.go]: One more note, you are modifying the slice here, no real need to return it. The swap can be done in one line
`slice[pos1], slice[pos2] = slice[pos2], slice[pos1]`
so it is debateable whether you need a separate function to do so, though it does make things slightly more readable in the sort functions.

File: 02_bubble/willshen8/bubble_test.go (hunk 7 above)
Comment (marksomething) [reply 02_bubble/willshen8/bubble_test.go]: You used different tests scenarios for bubble and insertionSort, it's not clear why they would not be the same.

File: 02_bubble/willshen8/bubble_test.go (hunk 8 above)
Comment (marksomething) [reply 02_bubble/willshen8/bubble_test.go]: It's a bit of a style choice I guess. The performance impact is trivial for this usage. I'm happy to accept this.

File: 02_bubble/willshen8/bubble.go (hunk 2 above)
Comment (marksomething) [reply 02_bubble/willshen8/bubble.go]: Don't forget to push your changes.

File: 02_bubble/willshen8/bubble_test.go (hunk 7 above)
Comment (willshen8) [reply 02_bubble/willshen8/bubble_test.go]: ok, thanks, I thought you meant upper or lower cases.

File: 02_bubble/willshen8/bubble.go (hunk 4 above)
Comment (willshen8) [reply 02_bubble/willshen8/bubble.go:25]: I couldn't get it to work without making my code confusing. I will definitely take on this advise in the future.

File: 02_bubble/willshen8/bubble.go (hunk 2 above)
Comment (willshen8) [reply 02_bubble/willshen8/bubble.go]: Thanks Dan, I copied the slice into a copySlice and only returning the copied one. So not mutating it.

File: 02_bubble/willshen8/bubble_test.go (hunk 7 above)
Comment (willshen8) [reply 02_bubble/willshen8/bubble_test.go]: Done, I included all test cases for both sort and merged the two test functions into one.

File: 02_bubble/willshen8/bubble_test.go (hunk 9 above)
Comment (anzboi) [reply 02_bubble/willshen8/bubble_test.go:38]: I find subtests are sometimes warranted and other times not.
Good: Different test cases (edge cases, different input types etc)
Bad: Randomized tests
In this case it would be a good idea to use them, but I will leave this in the 'do next time' bucket. since you do go through the effort to report the name in the test logs.
To run a subtest
```
t.Run(name, func(t *testing.T) {
  // sub test, written just like any other test, have access to test data
})
```

File: 02_bubble/willshen8/bubble_test.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+		{name: "Slice with one number different to rest", input: []int{1, 1, 1, 2, 1},
+			expected: []int{1, 1, 1, 1, 2}},
+		{name: "Slice with negative number", input: []int{1, -1}, expected: []int{-1, 1}},
+	}
+
+	for _, test := range tests {
+		actual := bubble(test.input)
+		testResult := equal(actual, test.expected)
+		if testResult == false {
+			t.Fatalf("Bubble Sort Test Failed! Expected: %v, Got: %v", test.expected, actual)
```
Comment (anzboi) [inline 02_bubble/willshen8/bubble_test.go:42]: I recommend use `Errorf` instead of `Fatalf`. Using `Fatalf` only causes you to see the first case that failed, not all the failing cases.

File: 02_bubble/willshen8/bubble_test.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+	}
+}
+
+func TestBubbleAndInsertionSort(t *testing.T) {
+	tests := []struct {
+		name     string
+		input    []int
+		expected []int
+	}{
+		{name: "Empty slice", input: []int{}, expected: []int{}},
```
Comment (anzboi) [inline 02_bubble/willshen8/bubble_test.go:28]: Great test coverage and incorporating both sorts in the same test. We will do suite testing when we get into the rest api and you will see how you can run the same test on multiple implementations of a particular function/interface.

**Review by anzboi — DISMISSED** [review]
LGTM.
I've commented on how you can run sub-tests, but don't feel the need to change for this PR, keep in mind for next time.

**Review by marksomething — DISMISSED** [review]
👍

**Review by anzboi — APPROVED** [review]
LGTM
